New Rule: 2.4.1 Bypass Blocks (cf77f2) (#421)
act-rules/act-rules.github.io: _rules/first-focusable-element-is-link-to-main-8a213c.md

    passed: satisfied
    inapplicable: further testing needed
input_aspects:
  - DOM Tree
  - CSS styling
  - Language
acknowledgments:
  authors:
    - Jean-Yves Moyen
    - Christina Adams
  assets:
    - _The Three Kingdoms_ by Yu Sumei (Tuttle publishing, May 2014, ISBN 9780804843935)
---

## Applicability

This rule applies to any [HTML web page][].

## Expectation 1

The first [focusable][] element within the test target:

- is [included in the accessibility tree][]; and
- is [visible][] when [focused][]; and
- has a [semantic role][] of `link`; and
- can be [activated][] by use of keyboard only; and
- when [activated][], moves focus [at the start][] of the [main section of content][] of the [document][]; and
- has an [accessible name][] that communicates that it links to the [main section of content][].

## Assumptions

- This rule assumes that description of the link is provided through its [accessible name][].
- This rule assumes that [Technique G1: Adding a link at the top of each page that goes directly to the main content area][tech g1] requires that the link can be [activated][] by use of keyboard only (in order to be useful for keyboard users).

> - This rule assumes that any global dismissible information that only appears once per site has already been acknowledged and is not displayed anymore. Many sites display a cookies policy banner which might be stealing focus until dismissed (usually by viewing and accepting cookies policy). Since that content is _not_ repeated (is it only shown once for the full site), it is not a problem to have it, and it may appear on any page of the site (depending where the user first comes in). If such a banner is taken into account, the rule may fail incorrectly.

Reviewer: I've had to read this several times until I remembered what bugged me. And it now seems not that relevant anymore. Still, I would like to suggest a simpler formulation (I'm just suggesting removing one sentence that just repeats information):

```suggestion
- This rule assumes that any global dismissible information that only appears once per site has already been acknowledged and is not displayed anymore. Many sites display a cookies policy banner which might be stealing focus until dismissed (usually by viewing and accepting cookies policy). If such a banner is taken into account, the rule may fail incorrectly.
```

---
id: 8a213c
name: First focusable element is link to main area of content
rule_type: atomic
description: |
  This rule checks that the first focusable element is a link to non-repeated content in the page
accessibility_requirements:
  wcag-technique:G1: # Adding a link at the top of each page that goes directly to the main content area
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed
input_aspects:
  - Accessibility Tree
  - CSS styling
  - DOM Tree
  - Language
acknowledgments:
  authors:
    - Christina Adams
    - Jean-Yves Moyen
  assets:
    - _The Three Kingdoms_ by Luo Guanzhong, translation by Yu Sumei (Tuttle publishing, 2014, ISBN 9780804843935)
---

## Applicability

This rule applies to any [HTML web page][].

## Expectation 1

> Each test target has a non-empty [sequential focus navigation order][sequential focus navigation].

Reviewer 2: This seems unnecessary. If this is false, so is the second expectation.

## Expectation 2

For each test target, all the following are true for the first element in its [sequential focus navigation order][sequential focus navigation]:

- the focusable element is [keyboard actionable][]; and
- the focusable element is [included in the accessibility tree][]; and
- the focusable element is a [semantic link][]; and

> - when the focusable element is [activated][], focus moves [at the end][] of the first [block of repeated content][] (in the [flat tree][]) with [perceivable content][] [at its end][at the end] which is not part of any [block of repeated content][]; and

Reviewer 2: This is pretty confusing, can you rephrase this in some way? I'm not sure I understand what this is saying.

---
id: 8a213c

> name: First focusable element is link to main content

Reviewer 3: Keeping this rule as Best Practice given that it was already written and that both Siteimprove and Qualweb have expressed interest in implementing it anyway.

rule_type: atomic
description: |
  This rule checks that the first focusable element is a link to the main content of the page
accessibility_requirements:
  wcag-technique:G1: # Adding a link at the top of each page that goes directly to the main content area
    forConformance: false
    failed: not satisfied
    passed: satisfied
    inapplicable: further testing needed
input_aspects:
  - Accessibility Tree
  - CSS styling
  - DOM Tree
  - Language
acknowledgments:
  authors:
    - Christina Adams
    - Jean-Yves Moyen
  assets:
    - _The Three Kingdoms_ by Luo Guanzhong, translation by Yu Sumei (Tuttle publishing, 2014, ISBN 9780804843935)
---

## Applicability

This rule applies to any [HTML web page][].

## Expectation 1

> There is at least one [focusable][] element within the test target.

Reviewer 4: There's a disconnect between this expectation and the next. Just because there's a focusable element "within the test target", which is ambiguously phrased, doesn't mean that the sequential focus navigation order, which is a sequence of elements, is non-empty. The expectation should probably be closer to:

> The sequential focus navigation order of the test target is non-empty.

> The [sequential focus navigation order][sequential focus navigation] of the test target is not empty.

Reviewer 2: ```suggestion
The [sequential focus navigation order][sequential focus navigation] of each test target is not empty.
```

## Expectation 2

> The first [focusable][] element within the test target:

Reviewer 4: The expectation doesn't define the ordering of focusable elements and so "first" is ambiguous. From the test cases, it seems like we want the first element in sequential navigation order.

> The first [focusable][] element (in [sequential focus navigation][] order) within the test target:

Reviewer 4: By definition, an element within the sequential focus navigation order has to be focusable.

> The first element in the [sequential focus navigation order][sequential focus navigation] of the test target:

Reviewer 2: ```suggestion
For each test target, the all following is true for the first element in the [sequential focus navigation order][sequential focus navigation]:
```

> - is [keyboard actionable][]; and
> - is [included in the accessibility tree][]; and
> - is a [semantic link][]; and
> - when [activated][], moves focus [at the start][] of the [main block of content][] of the [document][]; and
> - has an [accessible name][] that communicates that it links to the [main block of content][].

Reviewer 2: ```suggestion
- the focusable element is [keyboard actionable][]; and
- the focusable element is [included in the accessibility tree][]; and
- the focusable element is a [semantic link][]; and
- when the focusable element is [activated][], moves focus [at the start][] of the [main block of content][] of the [document][]; and
- the focusable element has an [accessible name][] that communicates that it links to the [main block of content][].
```

## Assumptions

- This rule assumes that there is exactly one [main block of content][] inside each [HTML web page][].
- This rule assumes that the description of the link is provided through its [accessible name][].
- This rule assumes that [Technique G1: Adding a link at the top of each page that goes directly to the main content area][tech g1] requires that the link can be [activated][] by use of keyboard, including being part of [sequential focus navigation][] (in order to be useful for keyboard users).
- This rule assumes that any global dismissible information that only appears once per site has already been acknowledged and is not displayed anymore. Many sites display a cookies policy banner which might be stealing focus until dismissed (usually by viewing and accepting cookies policy). If such a banner is taken into account, the rule may fail incorrectly.
- This rule assumes that the language of each test target can be correctly determined (either programmatically or by analyzing the content), and sufficiently understood.

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

> This rule and [Technique G1: Adding a link at the top of each page that goes directly to the main content area][tech g1] are best practices to satisfy [Success Criterion 2.4.1 Bypass blocks][sc241]. It is possible to satisfy it by other means. Moreover, any document passing this rule will also pass rule [Document has an instrument to move focus to main block of content][], therefore, this rule is not need to pass rule [Bypass blocks of content][bypass blocks].

Reviewer: ```suggestion
This rule and [Technique G1: Adding a link at the top of each page that goes directly to the main content area][tech g1] are best practices to satisfy [Success Criterion 2.4.1 Bypass blocks][sc241]. It is possible to satisfy it by other means. Moreover, any document passing this rule will also pass rule [Document has an instrument to move focus to main block of content][], therefore, this rule is not needed to pass rule [Bypass blocks of content][bypass blocks].
```

				of time.
			</p>
			<a href="/test-assets/bypass-blocks-cf77f2/chapter2.html">Read Chapter 2</a>
		</main>
	</body>
</html>
```

#### Passed Example 2

In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. In this case, the link is normally hidden but is [visible][] when [focused][].

```html
<html lang="en">
	<head>
		<link rel="stylesheet" href="../test-assets/bypass-blocks-cf77f2/styles.css" />
		<title>The Three Kingdoms, Chapter 1</title>
	</head>
	<body>
		<a href="#main" class="visible-on-focus">Skip to main content</a>

		<aside id="about-book">
			<h1>About the book</h1>
			<p>The Romance of the Three Kingdoms is a 14th century historical novel.</p>
		</aside>

		<main id="main">
			<h1>Three Heroes Swear Brotherhood at a Feast in the Peach Garden</h1>
			<p>
				Unity succeeds division and division follows unity. One is bound to be replaced by the other after a long span
				of time.
			</p>
			<a href="/test-assets/bypass-blocks-cf77f2/chapter2.html">Read Chapter 2</a>
		</main>
	</body>
</html>
```

#### Passed Example 3

> In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. In this case, the [activation][activated] behavior, and the possibility to [activate][activated] the element with keyboard, is done by scripting.

Reviewer: ```suggestion
In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. In this case, the [activation][activated] behavior, and the possibility to [activate][activated] the element with keyboard, is done by scripting and by having a `tabindex` attribute with a value of 0.
```

		</main>
	</body>
</html>
```

#### Passed Example 6

In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. In this case, it is the first [focusable][] element, even though it is the second link in tree order, but the focus order has been changed by the `tabindex` attributes.

```html
<html lang="en">
	<head>
		<title>The Three Kingdoms, Chapter 1</title>
	</head>
	<body>
		<nav>
			<a href="https://act-rules.github.io/" tabindex="2">ACT rules</a>
			<a href="#main" tabindex="1">Skip to main content</a>
		</nav>

		<aside id="about-book">
			<h1>About the book</h1>
			<p>The Romance of the Three Kingdoms is a 14th century historical novel.</p>
		</aside>

		<main id="main">
			<h1>Three Heroes Swear Brotherhood at a Feast in the Peach Garden</h1>
			<p>
				Unity succeeds division and division follows unity. One is bound to be replaced by the other after a long span
				of time.
			</p>
			<a href="/test-assets/bypass-blocks-cf77f2/chapter2.html">Read Chapter 2</a>
		</main>
	</body>
</html>
```

#### Passed Example 7

> In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. Even through its target is inside another [block of content][], it is still [at the start][] of the [main block of content][] because there is no [perceivable content][] between the link target and the [main block of content][]. Thus, following the link does skip all the repeated content.

Reviewer: ```suggestion
In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. Even though its target is inside another [block of content][], it is still [at the start][] of the [main block of content][] because there is no [perceivable content][] between the link target and the [main block of content][]. Thus, following the link does skip all the repeated content.
```

```html
<html lang="en">
	<head>
		<title>The Three Kingdoms, Chapter 1</title>
	</head>
	<body>
		<a href="#main">Skip to main content</a>

		<aside id="about-book">
			<h1>About the book</h1>
			<p>The Romance of the Three Kingdoms is a 14th century historical novel.</p>
			<span id="main"></span>
		</aside>

		<main>
			<h1>Three Heroes Swear Brotherhood at a Feast in the Peach Garden</h1>
			<p>
				Unity succeeds division and division follows unity. One is bound to be replaced by the other after a long span
				of time.
			</p>
			<a href="/test-assets/bypass-blocks-cf77f2/chapter2.html">Read Chapter 2</a>
		</main>
	</body>
</html>
```

#### Passed Example 8

> In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. Even through its target is not the first element in it, it is still [at the start][] of the [main block of content][] because it is before any [perceivable content][] before the target inside the [main block of content][]. Thus, following the link does not skip any non-repeated content.

Reviewer: ```suggestion
In this [document][], the first [focusable][] element is a [keyboard actionable][] skip link; it is [included in the accessibility tree][] and has a descriptive [accessible name][]. Even though its target is not the first element in it, it is still [at the start][] of the [main block of content][] because it is before any [perceivable content][] before the target inside the [main block of content][]. Thus, following the link does not skip any non-repeated content.
```

- has either an [accessible name](#accessible-name) or [accessible description](#accessible-description) that communicates that it links to the [main content](#main-content).

## Assumptions

This rule assumes that any global dismissible information that only appears once per site has already been acknowledged and is not displayed any more.

**Note**: The aim of such link is to be able to skip repeated content (headers, navigation bar, ...) when viewing several pages of the same site. Many sites display a cookies policy banner which might be stealing focus until dismissed (usually be viewing and accepting cookies policy). Since that content is _not_ repeated (is it only shown once for the full site), it is not a problem to have it, and it may appear on any page of the site (depending where the user first comes in).

## Accessibility Support

_There are no major accessibility support issues known for this rule._

## Background

- [G1: Adding a link at the top of each page that goes directly to the main content area](https://www.w3.org/WAI/WCAG21/Techniques/general/G1)

> ## Test Cases

Reviewer 3: Given that "main content" is going to be a slightly blurry term, I suggest adding to each example description "main content is identified here by the `main` element."

### Passed

#### Passed Example 1

The link to skp repeated content is visible, is included in the accessibility tree, and links to a valid HTML element.

```html

> <!DOCTYPE html>

Reviewer 3: In this rule, it is probably sufficient to have the content of the `body` as examples.

Reviewer 5: Do you mean to remove the extra document code in each example? Only leaving the body tag and contents?

Reviewer 3: Yes, even just the content of the `body` (not the `body` tag themselves) is enough. That's what we usually do (my bad for putting too big example templates in these rules…)

<html lang="en">
	<head>
		<title></title>
	</head>
	<body>
		<nav>
			<a href="#main-content">Skip to main content</a>
		</nav>

> 		<!-- Repeated Content -->

Reviewer 3: Maybe put the repeated content comment inside a specific element to make it more visible, _e.g._
```html
<nav>
  <a href="#main-content">Skip to main content</a>
</nav>
<div id="repeated-content">
  <!-- Repeated Content -->
</div>
<main id="main-content">
  Main Content
</main>
```

(on all examples)

Reviewer 5: Great suggestion. Will add to each example.

	<head>
		<title></title>
	</head>
	<body>
		<style>
			#skipNav a {
				height: 100px;
				position: absolute;
				top: -100px;
			}
			#skipNav a:focus {
				top: 0px;
			}
		</style>
		<nav id="skipNav">
			<a href="#main-content">Skip to main content</a>
		</nav>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

#### Passed Example 3

The link to skip repeated content has an `aria-label` to provide an accessible name.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title></title>
	</head>
	<body>
		<nav>
			<a href="#main-content" aria-label="Skip to main content"></a>
		</nav>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

### Failed

#### Failed Example 1

There is no link to skip repeated content.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title></title>
	</head>
	<body>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

#### Failed Example 2

The link to skip repeated content is not the first focusable element within the `body`.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title></title>
	</head>
	<body>

> 		<a href="/anotherpage">A link to another page</a>

Reviewer 3: Putting a dead link in examples is not so good. Can easily be replaced with a link to WCAG, W3C, or whatever other cool resource 😎

Reviewer 5: Sorry! I will add a link to a real external resource 😄

	</body>
</html>
```

#### Failed Example 3

The link to skip repeated content does not have an accessible name.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title></title>
	</head>
	<body>
		<nav>
			<a href="#main-content"></a>
		</nav>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

#### Failed Example 4

The link to skip repeated content does not reference a valid `id` and will not move focus to the main content when activated.

```html
<!DOCTYPE html>
<html lang="en">
	<head>
		<title></title>
	</head>
	<body>
		<nav>
			<a href="#invalidId">Skip to main content</a>
		</nav>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

#### Failed Example 5

The link to skip repeated content is not visible, and is not included in the accessibility tree.

```html
<!DOCTYPE html>
<html lang="eng">
	<head>
		<title></title>
	</head>
	<body>
		<nav>
			<a href="#main-content" style="display:none;">Skip to main content</a>
		</nav>
		<!-- Repeated content -->
		<main id="main-content">Main Content</main>
	</body>
</html>
```

#### Failed Example 6

> The link to skip repeated content does not have a semantic role of link.

Reviewer 3: "The **link** to skip" is not really correct since it is, well, not a link…

And I think it would be a better example if it was actually skipping to main content (_e.g._ a button or something with a correct `onClick` property). So that we show that the problem is "not being a link" rather than "not moving focus to the correct place".

Reviewer 5: Good catch. Wasn't sure how to describe. I will take another shot at this one and add the click event functionality
